tutorials/examples/train_graph_ring.py

```diff
@@ -0,0 +1,743 @@
+            action_type = self.action_norm(action_type)
+
+        action_type = self._group_mean(
+            torch.mean(action_type, dim=-1, keepdim=True), batch_ptr
+        )
+
+        # Multiple action type convolutions with residual connections.
+        edge_feature = x
+        for i in range(0, len(self.edge_conv_blks), 4):
+
+            # GIN/GCN conv.
+            edge_feature_new = self.edge_conv_blks[i](edge_feature, edge_index.T)
+            # First linear.
+            edge_feature_new = self.edge_conv_blks[i + 1](edge_feature_new)
+            # ReLU.
+            edge_feature_new = self.edge_conv_blks[i + 2](edge_feature_new)
+            # Second linear.
+            edge_feature_new = self.edge_conv_blks[i + 3](edge_feature_new)
+            # Residual connection with original input.
+            edge_feature = edge_feature_new + edge_feature
+            edge_feature = self.edge_norm(edge_feature)
+
+        # edge_feature = self._group_mean(
+        #     torch.mean(edge_feature, dim=-1, keepdim=True), batch_ptr
+        # )
+
+        edge_feature = edge_feature.reshape(
+            *states_tensor["batch_shape"], self.n_nodes, self.hidden_dim
+        )
+
+        # This is n_nodes ** 2, for each graph.
+        edge_index = torch.einsum("bnf,bmf->bnm", edge_feature, edge_feature)
+
+        # Undirected.
+        if self.is_directed:
+            i_up, j_up = torch.triu_indices(
+                self.n_nodes, self.n_nodes, offset=1
+            )  # Upper triangle.
+            i_lo, j_lo = torch.tril_indices(
+                self.n_nodes, self.n_nodes, offset=-1
+            )  # Lower triangle.
+
+            # Combine them
+            i0 = torch.cat([i_up, i_lo])
+            i1 = torch.cat([j_up, j_lo])
+            out_size = self.n_nodes**2 - self.n_nodes
+
+        else:
+            i0, i1 = torch.triu_indices(self.n_nodes, self.n_nodes, offset=1)
+            out_size = (self.n_nodes**2 - self.n_nodes) // 2
+
+        # Grab the needed elems from the adjacency matrix and reshape.
+        batch_arange = torch.arange(batch_size)
+        edge_actions = edge_index[batch_arange[:, None, None], i0, i1]
+        edge_actions = edge_actions.reshape(*states_tensor["batch_shape"], out_size)
+
+        if self.is_backward:
+            return edge_actions
+        else:
+            return torch.cat([edge_actions, action_type], dim=-1)
+
+
+class RingGraphBuilding(GraphBuilding):
+    """Override the GraphBuilding class to create have discrete actions.
+
+    Specifically, at initialization, we have n nodes.
+    The policy can only add edges between existing nodes or use the exit action.
+    The action space is thus discrete and of size n^2 + 1, where the last action is the exit action,
+    and the first n^2 actions are the possible edges.
+
+    Args:
+        n_nodes: The number of nodes in the graph.
+    """
+
+    def __init__(self, n_nodes: int, state_evaluator: callable, directed: bool):
```

Comment:
for me, it doesn't make sense to subclass GraphBuilding without using it's init function

Reply:
my bad